// theme-ui/dist/index.js
var core = require('@theme-ui/core');
var colorModes = require('@theme-ui/color-modes');
var mdx = require('@theme-ui/mdx');
var themeProvider = require('@theme-ui/theme-provider');
var components = require('@theme-ui/components');
var css = require('@theme-ui/css');

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var BaseStyles = function BaseStyles(props) {
  return core.jsx('div', _extends({}, props, {
    sx: {
      fontFamily: 'body',
      lineHeight: 'body',
      fontWeight: 'body',
      variant: 'styles'
    }
  }));
};

Object.keys(components).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return components[k];
    }
  });
});
Object.defineProperty(exports, 'Context', {
  enumerable: true,
  get: function () {
    return core.Context;
  }
});
Object.defineProperty(exports, 'jsx', {
  enumerable: true,
  get: function () {
    return core.jsx;
  }
});
Object.defineProperty(exports, 'merge', {
  enumerable: true,
  get: function () {
    return core.merge;
  }
});
Object.defineProperty(exports, 'useThemeUI', {
  enumerable: true,
  get: function () {
    return core.useThemeUI;
  }
});
Object.defineProperty(exports, 'InitializeColorMode', {
  enumerable: true,
  get: function () {
    return colorModes.InitializeColorMode;
  }
});
Object.defineProperty(exports, 'useColorMode', {
  enumerable: true,
  get: function () {
    return colorModes.useColorMode;
  }
});
Object.defineProperty(exports, 'Styled', {
  enumerable: true,
  get: function () {
    return mdx.Styled;
  }
});
Object.defineProperty(exports, 'components', {
  enumerable: true,
  get: function () {
    return mdx.components;
  }
});
Object.defineProperty(exports, 'ThemeProvider', {
  enumerable: true,
  get: function () {
    return themeProvider.ThemeProvider;
  }
});
Object.defineProperty(exports, 'css', {
  enumerable: true,
  get: function () {
    return css.css;
  }
});
Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function () {
    return css.get;
  }
});
exports.BaseStyles = BaseStyles;
